railties/lib/rails/generators/rails/app/templates/.devcontainer/docker-compose.yml.tt

```diff
@@ -0,0 +1,78 @@
+services:
+  rails-app:
+    build:
+      context: ..
+      dockerfile: .devcontainer/Dockerfile
+
+    volumes:
+      - ../..:/workspaces:cached
+
+    # Overrides default command so things don't shut down after the process ends.
+    command: sleep infinity
+
+    networks:
+      - default
+
+    # Uncomment the next line to use a non-root user for all processes.
+    # user: vscode
+
+    # Use "forwardPorts" in **devcontainer.json** to forward an app port locally.
+    # (Adding the "ports" property to this file will not forward from a Codespace.)
+    ports:
+      - 45678:45678
+<%- if !devcontainer_dependencies.empty? -%>
+    depends_on:
+<%- devcontainer_dependencies.each do |dependency| -%>
+      - <%= dependency %>
+<%- end -%>
+<%- end -%>
+
+<%- if depends_on_system_test? -%>
+  selenium:
+    image: seleniarm/standalone-chromium
+    restart: unless-stopped
+    networks:
+      - default
+<%- end -%>
+
+<%- unless options.skip_active_job? && options.skip_action_cable? -%>
+  redis:
+    image: redis:7.2
+    restart: unless-stopped
+    networks:
+      - default
+    ports:
+      - 6379:6379
+<%- end -%>
+
+<%- if options.database == "postgresql" -%>
+  postgres:
+    image: postgres:16.1
+    restart: unless-stopped
+    networks:
+      - default
+    environment:
+      POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: postgres
+<%- end -%>
+
+<%- if options.database == "mysql" -%>
+  mysql:
+    image: mysql/mysql-server:8.0
+    restart: unless-stopped
+    environment:
+      MYSQL_ALLOW_EMPTY_PASSWORD: true
+      MYSQL_ROOT_HOST: "%"
+    networks:
+      - default
+<%- end -%>
+
+<%- if options.database == "trilogy" -%>
+  mariadb:
+    image: mariadb:10.5
+    restart: unless-stopped
+    networks:
+      - default
+    environment:
+      MARIADB_ALLOW_EMPTY_ROOT_PASSWORD: true
+<%- end -%>
```

Comment:
Moving all services in `devcontainer_dependencies` (`depends_on`) to a `docker-compose.yml.tt` root folder file, would improve dev environment and adoption to docker.

Because, having all service dependencies in root `docker-compose.yml.tt` allow us to start the services only with `docker compose` and the Rails server locally.  Meaning, along with dev containers, we could use service with `docker compose up -d`, so every developer could choose between:

- local server and local services
- dev containers in Docker
- local server and services in Docker

without changing any line in settings.

Reply:
The goal of this work is to create a light weight development environment that is easy to get working without installing many dependencies, especially for people new to Rails and even new to coding. While I can definitely see the value of having the Rails app running locally and other services in the container, it is a more advance setup and not exactly with the scope of our goals for this work. So I am not sure we want to make that a default part of the configuration and thus something the Rails team has to support going forward.

Rafael is away for a couple days, let's get his opinion when he is back.

Reply:
As you can see in RubyGems repo, there are no more dependencies, and it is actually more `light weight development environment` as going with Dev Containers. In case of RubyGems, I was working on a feature there, I was not able to run Dev Containers, but I could easily run `docker compose up` and working on this project.

Would love to hear Rafael's opinion too.

Reply:
You can still use `docker compose up -f .devcontainer/docker-compose.yml -d <service>`.

Having dev env configuration in a dedicated folder (like `.devcontainer`) would prevent from potential conflicts with production `docker-compose.yml` (which is usually kept in the root) or other Docker configurations.

Reply:
Yes, this sounds good to me too.